services/liquidity.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  SigningCosmWasmClient,
+  CosmWasmClient,
+  MsgExecuteContractEncodeObject,
+} from '@cosmjs/cosmwasm-stargate'
+import { MsgExecuteContract } from '@cosmjs/cosmwasm-stargate/build/codec/cosmwasm/wasm/v1beta1/tx'
+import { toUtf8 } from '@cosmjs/encoding'
+import { coin, StdFee } from '@cosmjs/launchpad'
+import { BroadcastTxResponse } from '@cosmjs/stargate'
+
+export type AddLiquidityInput = {
+  nativeAmount: number
+  nativeDenom: string
+  maxToken: number
+  minLiquidity: number
+  senderAddress: string
+  swapAddress: string
+  tokenAddress: string
+  client: SigningCosmWasmClient
+}
+
+export const addLiquidity = async (
+  input: AddLiquidityInput
+): Promise<BroadcastTxResponse> => {
+  let msg1 = {
+    increase_allowance: {
+      amount: `${input.maxToken}`,
+      spender: `${input.swapAddress}`,
+    },
+  }
+  const executeContractMsg1: MsgExecuteContractEncodeObject = {
+    typeUrl: '/cosmwasm.wasm.v1beta1.MsgExecuteContract',
+    value: MsgExecuteContract.fromPartial({
+      sender: input.senderAddress,
+      contract: input.tokenAddress,
+      msg: toUtf8(JSON.stringify(msg1)),
+      funds: [],
+    }),
+  }
+  let msg2 = {
+    add_liquidity: {
+      max_token: `${input.maxToken}`,
+      min_liquidity: `${input.minLiquidity}`,
+    },
+  }
+  const executeContractMsg2: MsgExecuteContractEncodeObject = {
+    typeUrl: '/cosmwasm.wasm.v1beta1.MsgExecuteContract',
+    value: MsgExecuteContract.fromPartial({
+      sender: input.senderAddress,
+      contract: input.swapAddress,
+      msg: toUtf8(JSON.stringify(msg2)),
+      funds: [coin(input.nativeAmount, input.nativeDenom)],
+    }),
+  }
+  const fee: StdFee = {
+    amount: input.client.fees.exec.amount,
```

Comment:
curious if this should be multiplied by 2 as well

Reply:
most likely. we will need to actually figure out the appropriate gas fee at some point